File: apps/api/src/card_in_repo_api/concepts.py
from __future__ import annotations

from typing import Any

from .teaching import build_basic_explanation, evidence_id


def build_concept_candidates(facts: dict[str, Any], features: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Derive teachable concept candidates from static-analysis evidence only."""
    symbols = {symbol["id"]: symbol for symbol in facts.get("symbols", []) if symbol.get("id")}
    symbol_paths = facts.get("symbol_paths", {})
    concepts: list[dict[str, Any]] = []
    seen: set[str] = set()

    for feature in features:
        steps = feature.get("flow_steps", [])
        if not steps:
            continue
        concept_id = f"concept:flow:{feature['id']}"
        evidence: list[dict[str, Any]] = []
        for index, step in enumerate(steps):
            symbol = symbols.get(step.get("symbol_id"))
            if not symbol:
                continue
            evidence.append({
                "id": evidence_id(concept_id, index),
                "symbol_id": symbol["id"],
                "symbol_name": symbol.get("name"),
                "path": symbol_paths.get(symbol["id"]),
                "range": symbol.get("range"),
                "relation": step.get("relation"),
                "order": step.get("order"),
            })
        if not evidence:
            continue
        if concept_id in seen:
            continue
        seen.add(concept_id)
        concept = {
            "id": concept_id,
            "kind": "execution_flow",
            "name": feature.get("name") or evidence[0].get("symbol_name") or "Execution flow",
            "feature_id": feature["id"],
            "evidence": evidence,
        }
        concept["explanation"] = build_basic_explanation(concept)
        concepts.append(concept)
    return concepts

File: apps/web/src/concepts.ts
import type { Feature, RepositoryFile } from './api';

export type ConceptEvidence = { symbol_id:string; symbol_name:string; path?:string; relation?:string; order:number };
export type ConceptCandidate = { id:string; name:string; kind:'execution_flow'; feature_id:string; evidence:ConceptEvidence[] };

export function buildConceptCandidates(features: Feature[], files: RepositoryFile[]): ConceptCandidate[] {
  const paths = new Map(files.flatMap(file => file.symbols.map(symbol => [symbol.id, file.path] as const)));
  return features.flatMap(feature => {
    const evidence = feature.flow_steps.map((step,index) => ({
      symbol_id: step.symbol_id,
      symbol_name: step.symbol_name ?? step.symbol_id,
      path: paths.get(step.symbol_id),
      relation: step.relation,
      order: step.order ?? index + 1,
    }));
    if (!evidence.length) return [];
    return [{id:`concept:flow:${feature.id}`,name:feature.name,kind:'execution_flow' as const,feature_id:feature.id,evidence}];
  });
}
